elfi/methods/diagnostics.py

```diff
@@ -0,0 +1,294 @@
+"""Methods for ABC diagnostics."""
+
+import logging
+from itertools import combinations
+
+import numpy as np
+from scipy.spatial import cKDTree
+from scipy.special import digamma, gamma
+
+import elfi
+
+logger = logging.getLogger(__name__)
+
+
+class TwoStageSelection:
+    """Perform the summary-statistics selection proposed by Nunes and Balding (2010).
+
+    The user can provide a list of summary statistics as list_ss, and let ELFI to combine them,
+    or provide some already combined summary statistics as prepared_ss.
+
+    The rationale of the Two Stage procedure procedure is the following:
+    - First, the module computes or accepts the combinations of the candidate summary statistics;
+    - In Stage 1, each summary-statistics combination is evaluated using the
+    Minimum Entropy algorithm;
+    - In Stage 2, the minimum-entropy combination is selected,
+    and the `closest' datasets are identified;
+    - Further in Stage 2, for each summary-statistics combination,
+    the mean root sum of squared errors (MRSSE) is calculated over all `closest datasets',
+    and the minimum-MRSSE combination is chosen as the one with the optimal performance.
+
+    Attributes
+    ----------
+    simulator : elfi.Node
+        Node (often elfi.Simulator) for which the summary statistics will be applied.
+        The node is the final node of a coherent ElfiModel (i.e. it has no child nodes).
+    fn_distance : str or callable function
+        Function for calculating distance (can be a string if the function is implemented in ELFI).
+    list_ss : List of callable functions, optional
+        List of uncombined candidate summary statistics.
+    prepared_ss : List of lists of callable functions, optional
+        List of combined candidate summary statistics.
+    max_cardinality : int, optional
+        Maximum cardinality of a candidate summary-statistics combination.
+    seed : int, optional
+
+    References
+    ----------
+    [1] Nunes, M. A., & Balding, D. J. (2010).
+    On optimal selection of summary statistics for approximate Bayesian computation.
+    Statistical applications in genetics and molecular biology, 9(1).
+    [2] Blum, M. G., Nunes, M. A., Prangle, D., & Sisson, S. A. (2013).
+    A comparative review of dimension reduction methods in approximate Bayesian computation.
+    Statistical Science, 28(2), 189-208.
+
+    # TODO: include different distance functions in the summary-statistics combinations.
+
+    """
+
+    def __init__(self, simulator, fn_distance, list_ss=None, prepared_ss=None,
+                 max_cardinality=4, seed=0):
+        """Initialise the summary-statistics selection for the Two Stage Procedure.
+
+        Parameters
```

Comment:
Ah, actually better to just have either Attributes for the class or Parameters for the constructor, and I think ELFI tends to favour Parameters. I.e. no need to document the (identical) attributes above.

Otherwise you'll get the API documentation repeated.

Reply:
Removed the attribute documentation.